src/vmm/benches/queue.rs

```diff
@@ -0,0 +1,193 @@
+// Copyright 2024 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+//
+// Benchmarking cases:
+//   * `Queue.pop`
+//   * `Queue.add_used`
+//   * `DescriptorChain.next_descriptor`
+
+use std::num::Wrapping;
+
+use criterion::{criterion_group, criterion_main, Criterion};
+use vm_memory::GuestAddress;
+use vmm::devices::virtio::queue::{VIRTQ_DESC_F_NEXT, VIRTQ_DESC_F_WRITE};
+use vmm::devices::virtio::test_utils::VirtQueue;
+use vmm::utilities::test_utils::single_region_mem;
+
+/// Create one chain with n descriptors
+/// Descriptor buffers will leave with offest of 2048 bytes
+/// to leave some room for queue objects themselfs.
+/// Even though we don't really care about size of descriptors
+/// the maximum size of the packet is 64K and we are
+/// building a chain that has this length. For this
+/// we need each descriptor to be 64K / n size.
+fn set_dtable_one_chain(rxq: &VirtQueue, n: usize) {
+    for i in 0..n {
+        rxq.dtable[i].set(
+            (2048 + (65562 / n) * i) as u64,
+            (65562 / n) as u32,
+            VIRTQ_DESC_F_WRITE | VIRTQ_DESC_F_NEXT,
+            (i + 1) as u16,
+        );
+    }
+    rxq.dtable[n - 1].flags.set(VIRTQ_DESC_F_WRITE);
+    rxq.dtable[n - 1].next.set(0);
+    rxq.avail.ring[0].set(0);
+    rxq.avail.idx.set(n as u16);
+}
+
+/// Create n chains with 1 descriptors each
+/// Descriptor buffers will leave with offest of 2048 bytes
+/// to leave some room for queue objects themselfs.
+/// Even though we don't really care about size of descriptors
+/// the maximum size of the packet is 64K and we are
+/// building a set of descriptors with this total size.
+/// For this we need each descriptor to be 64K / n size.
+fn set_dtable_many_chains(rxq: &VirtQueue, n: usize) {
+    for i in 0..n {
+        rxq.dtable[i].set(
+            (2048 + (65562 / n) * i) as u64,
+            (65562 / n) as u32,
+            VIRTQ_DESC_F_WRITE,
+            0,
+        );
+        rxq.avail.ring[i].set(i as u16);
+    }
+    rxq.avail.idx.set(n as u16);
+}
+
+pub fn queue_benchmark(c: &mut Criterion) {
+    let mem = single_region_mem(2 * 65562);
+    let rxq = VirtQueue::new(GuestAddress(0), &mem, 256);
+    let mut queue = rxq.create_queue();
+
+    set_dtable_one_chain(&rxq, 1);
+    c.bench_function("next_descriptor_1", |b| {
+        b.iter(|| {
+            queue.next_avail = Wrapping(0);
+            // SAFETY: queue has 1 desc chanin
+            let desc = unsafe { queue.pop(&mem).unwrap_unchecked() };
+            let mut head = Some(desc);
+            while let Some(d) = head {
+                head = d.next_descriptor();
+            }
```

Comment:
Given that we are increasingly using `IoVecBuffer(Mut)` objects to parse these, maybe it would make sense to use them here?

Reply:
this is a bench just for `next_descriptor` call. Are saying we need another bench for parsing of `IoVecBuffer(Mut)` from desc chains?

Reply:
Not sure, I think it's fine now. Let's get this one merged and we'll think about `IoVecBuffer` stuff later.